### PR TITLE
mbedtls: add CONFIG_MBEDTLS_HEAP_SELF_INIT

### PR DESCRIPTION
**Pull request by daldridge-quic**
Add a config flag to enable conditional mbedtls heap
initialization at startup, defaulting to enabled.

Also make the mbedtls_init() non-static so it can be
called externally.

Signed-off-by: Eugene Cohen <quic_egmc@quicinc.com>
Signed-off-by: Dave Aldridge <quic_daldridg@quicinc.com>

### CODE REVIEW

**Comment by daldridge-quic** [comment]
@ceolin @d3zd3z @nashif @andyross is there anything i can do to help out with the review ?

File: include/zephyr/init.h
Comment (ceolin) [inline include/zephyr/init.h:180]: I don't think that is the right place to add it. Maybe would be better a header like `modules/mbedtls/zephyr_mbedtls.h` that you include to the include path when mbedTLS is enabled.

Comment (daldridge-quic) [reply include/zephyr/init.h:180]: Thanks for feedback, let me go noodle in the patch and see what i can come up with.

File: modules/mbedtls/CMakeLists.txt
Comment (mniestroj) [inline modules/mbedtls/CMakeLists.txt:12]: I think it could be just `include`, so that other headers would fit into that directory, like here: https://github.com/zephyrproject-rtos/zephyr/pull/46572/commits/30f07ed81f9f68d4c792ba0b062fc4d57a4a8589

File: modules/mbedtls/Kconfig
Comment (mniestroj) [inline modules/mbedtls/Kconfig:154]: Maybe just `config MBEDTLS_INIT_HEAP`?

File: modules/mbedtls/mbedtls_init/mbedtls_init.h
Comment (mniestroj) [inline modules/mbedtls/mbedtls_init/mbedtls_init.h:13]: Exposing `const struct device *device` does not seem right. I think we should have dedicated function like `zephyr_mbedtls_heap_init(void)` or something similar called from static `mbedtls_init(const struct device *device)` when proper Kconfig option is selected.

File: modules/mbedtls/CMakeLists.txt
Comment (daldridge-quic) [reply modules/mbedtls/CMakeLists.txt:12]: i shall go take a look thanks

File: modules/mbedtls/Kconfig
Comment (daldridge-quic) [reply modules/mbedtls/Kconfig:154]: sure

File: modules/mbedtls/mbedtls_init/mbedtls_init.h
Comment (daldridge-quic) [reply modules/mbedtls/mbedtls_init/mbedtls_init.h:13]: let me go have a noodle, thanks

**Review by tejlmand — CHANGES_REQUESTED** [review]
Exactly why is this PR needed ?

Could you try to explain what the problem is or open an issue.

Zephyr already supports:
https://github.com/zephyrproject-rtos/zephyr/blob/f3017aa787e75c77bb7deb9f4ae45d4b4e93fbc3/modules/mbedtls/Kconfig#L129-L130

where in the help text it even states:
https://github.com/zephyrproject-rtos/zephyr/blob/f3017aa787e75c77bb7deb9f4ae45d4b4e93fbc3/modules/mbedtls/Kconfig#L135-L136

As I see it, users who needs customization can already do so by disabling `MBEDTLS_ENABLE_HEAP=n` and then provide their own heap implementation, and initialize the heap whenever they want to.

I don't say this PR is not valuable, what i'm trying to understand is why the existing possibility of configuring and use mbedTLS is not sufficient in this case ?

File: modules/mbedtls/Kconfig
Comment (daldridge-quic) [reply modules/mbedtls/Kconfig:159]: Last things first, your suggested wording is great, thanks.
As to why it is needed... It is catering for a memory map where mbedTLS is separated out into its own area of rom. This area of rom exists in emulated environments but not in real hardware. We wanted to use the existing zephyr heap implementation for mbedTLS, just be in control of when it is called. This is perhaps a slightly unusual setup, but we figured it may well be useful for other users hence the PR. If the feeling is that this is too specialised a use case then no worries.

**Review by dcpleung — DISMISSED** [review]

File: modules/mbedtls/zephyr_init.c
Comment (mniestroj) [inline modules/mbedtls/zephyr_init.c:107]: `_mbedtls_init()` does more than just heap initialization. It also configured debug threshold (after rebase).

**Review by mniestroj — CHANGES_REQUESTED** [review]

File: modules/mbedtls/zephyr_init.c
Comment (daldridge-quic) [reply modules/mbedtls/zephyr_init.c:107]: Yes indeed, sorry I overlooked that, let me fix that right now.

**Comment by daldridge-quic** [comment]
um, didn't realise requesting a review would blow away existing requests, sorry about that @d3zd3z @nashif @andyross @tbursztyka @stephanosio

**Review by mniestroj — DISMISSED** [review]

File: modules/mbedtls/Kconfig
Comment (tejlmand) [reply modules/mbedtls/Kconfig]: thanks for the explanation, it's always helpful to understand the reason for a given patch.

I will let it be up to the maintainers to decide if this is valid additional ability in addition to existing use-cases.

File: modules/mbedtls/Kconfig
Comment (tejlmand) [inline modules/mbedtls/Kconfig:206]: No reason for `automatically` here.

If something is initialized at boot, then it must be kind of automatically.
```suggestion
	bool "Initialize mbed TLS at boot"
```

File: modules/mbedtls/Kconfig
Comment (tejlmand) [inline modules/mbedtls/Kconfig:208]: I believe this should be dependent upon `MBEDTLS_ENABLE_HEAP`

If there is no heap enabling in Zephyr mbedTLS, then this call is empty:
https://github.com/zephyrproject-rtos/zephyr/blob/4da89756b3b5853ec4773a20eec814c7af9062a4/modules/mbedtls/zephyr_init.c#L88
as can be seen here:
https://github.com/zephyrproject-rtos/zephyr/blob/4da89756b3b5853ec4773a20eec814c7af9062a4/modules/mbedtls/zephyr_init.c#L44-L46

Ref:
https://github.com/zephyrproject-rtos/zephyr/blob/79fccbb9426385c80d5c1001a3f612eeff4cd625/modules/mbedtls/Kconfig#L180-L181

Please reason if you disagree that `MBEDTLS_INIT` should not have this dependency.

**Review by tejlmand — CHANGES_REQUESTED** [review]
Some comments regarding the Kconfig.

Would like @d3zd3z or @ceolin to consider if this is valid use-case the changes proposed:
https://github.com/zephyrproject-rtos/zephyr/pull/47686#discussion_r947693747

Rest of changes looks fine to me.

**Review by d3zd3z — DISMISSED** [review]
I'm ok with this, once the issues mentioned are addressed.

File: modules/mbedtls/Kconfig
Comment (daldridge-quic) [reply modules/mbedtls/Kconfig:206]: agreed, i'll sort it

File: modules/mbedtls/Kconfig
Comment (daldridge-quic) [reply modules/mbedtls/Kconfig:208]: _mbedtls_init() does more than just init the heap, it can now also call mbedtls_debug_set_threshold(). Therefore, _mbedtls_init() needs to be called whether we are using the heap or not. So MBEDTLS_INIT doesn't need to be dependant on MBEDTLS_ENABLE_HEAP.

File: modules/mbedtls/Kconfig
Comment (daldridge-quic) [reply modules/mbedtls/Kconfig:205]: sure

**Comment by daldridge-quic** [comment]
@tejlmand I have addressed your feedback, if you could take a look that would be great.

**Comment by daldridge-quic** [comment]
@tejlmand @d3zd3z I believe I have addressed all feedback in the pull request. @andyross @dcpleung any chance of a review

**Review by dcpleung — APPROVED** [review]

**Review by mniestroj — APPROVED** [review]

**Comment by daldridge-quic** [comment]
@mniestroj @dcpleung thanks for the approvals.
@tejlmand i believe i have addressed all of your requested changes, could you please re-review.
@andyross any chance of a review please.

**Comment by daldridge-quic** [comment]
@d3zd3z As far as I can tell I have actioned all requested changes and I'm just waiting for review/approval. Can you please help here ?

**Comment by stephanosio** [comment]
@tejlmand FYI, your change request is still pending.
@daldridge-quic We are currently in a feature merge freeze for the upcoming 3.3 release. Unless you can justify this as fixing a bug, this will need to wait until 3.3 is released.

**Comment by daldridge-quic** [comment]
@stephanosio @tejlmand Now the v3.3.0 release has been made can this PR be addressed please.

**Comment by daldridge-quic** [comment]
@d3zd3z As far as I can tell I have actioned all requested changes and I'm just waiting for review/approval. Can you please help here ?

**Comment by daldridge-quic** [comment]
@stephanosio Hiya, you removed the `Stale` label from this MR last time it was applied, are you able to help again please ?

**Comment by fabiobaltieri** [comment]
@d3zd3z can you take another look?

**Review by tejlmand — APPROVED** [review]

**Comment by daldridge-quic** [comment]
Many thanks to everyone involved with getting this PR merged.

**Pull request merged**
